### ISSUE TITLE
Formatting paragraph can result in invalid syntax

### ISSUE THREAD

**Issue by mtzguido** [bohlender/vim-z3-smt2#4]
Hi! First of all thanks a lot for this extension, I'm finding it extremely useful.

A slight problem I usually run into is that my smt2 files have empty lines within a single top-level expression, such as this:
```smt2
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not
 
(= (+ x y) (+ y x))

))

(check-sat)
```
Placing the cursor at the beginning of the assert and formatting gives this, which is broken.
```smt2
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not (= (+ x y) (+ y x))))

(= (+ x y) (+ y x))

))

(check-sat)
```
So I now usually just remove all empty lines before working on a file, but if this worked out of the box it would be amazing :).

Come to think of it, the current behavior seems useful for formatting a whole chunk of s-expressions clumped together as a paragraph.. maybe there can be a binding to ask for formatting a single s-expression?

**Comment by bohlender** [comment]
I'm glad you're finding it helpful, and thanks for reporting the issue :+1: 

I've tried to reproduce the unexpected behaviour:
[![asciicast](https://asciinema.org/a/rOSUXFl6huTONR4SPi5GrrQf9.svg)](https://asciinema.org/a/rOSUXFl6huTONR4SPi5GrrQf9)
Formatting the current buffer works as expected but formatting the current paragraph yields a slightly different result.

However, in contrast to your report -- where we see `(= (+ x y) (+ y x))` twice -- I'm getting this:
```
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not (= (+ x y) (+ y x))))

))

(check-sat)
```
What Vim version are you at? I wonder what _exactly_ the bug is you wanted to report:
1. does formatting the paragraph really duplicate the expression `(= (+ x y) (+ y x))` for you,
1. or was that a typo and you rather wanted to report the empty line 5 that does not appear when we format the whole buffer?

> maybe there can be a binding to ask for formatting a single s-expression?

Sure, if you see need for it. Feel free to open a separate issue and give some examples of the expected behaviour. In particular, what shall happen when you try to format an already nested expression?

**Comment by mtzguido** [comment]
Hi Dimitri, thanks for replying. I seem to have left a whitespace in the middle line of the snippet I pasted... so it's not a completely "empty" line. Trying that exact snippet, I get the same behavior you show in the video, sorry about that (though it also shows a problem, as the result is syntactically invalid).

With this one, though, I get what I wrote before:
```smt2
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not

(= (+ x y) (+ y x))

))

(check-sat)
```
goes into
```smt2
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not (= (+ x y) (+ y x))))

(= (+ x y) (+ y x))

))

(check-sat)
```

So formatting really does duplicate the full expression for me, which makes the file syntactically invalid. The same happens in your video due to the extra `))`. I'm not sure I understood your (2) point but I don't particularly mind about empty lines, except for this issue of course :).

Vim version is the current one for debian testing
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Apr 03 2023 07:45:49)
Included patches: 1-3995, 4563, 4646, 4774, 4895, 4899, 4901, 4919
```

**Comment by bohlender** [comment]
With the new snippet I can reproduce the problematic duplication. I'll look into it.

**Comment by bohlender** [comment]
FYI: It turns out that the formatted string is not guaranteed to _replace_ the unformatted one in case of "paragraphs" in the s-expression. In your example the (proper) formatting only replaced line 4 but failed to delete the remnant lines 5-8.

The issue is not even in the parser or the formatter but in the subsequent replacement operation which naively relies on `}` to find the end of a "paragraph". It should rather rely on the source positions used by the parser.

**Comment by bohlender** [comment]
@mtzguido: I think the new branch fixes the issue. To give it a try just go to `~/.vim/bundle/vim-smt2` and check out `fix/format-paragraph-with-empty-lines` instead of `master`.

Let me know what you think about it. I've added the possibility to change what is formatted on press of `<leader>f`. To format the outermost S-expression simply add `nmap <leader>f <Plug>Smt2FormatOutermostSExpr` to your `~/.vimrc`. [It should handle even funky cases with multiple S-expressions in a line, such as  the following](https://asciinema.org/a/PK2oumNBFneihJWSfxtdHmNpD):
```smt
(declare-fun x () Int)
(declare-fun y () Int)

(assert (not false)) (assert (not

(= (+ x y)
   y)

)) (assert (> x y))

(check-sat
)
```

**Comment by mtzguido** [comment]
Hi @bohlender, thanks! This works a lot better for me. I did find one small quirk though, with something like:
```smt2
(assert (=

(+ x y)

5))
(assert (=

(+ x y)

5))
```
Using `<leader>f` on the second assertion will not format properly format either, I just get an extra line break.
```smt2
(assert (=

(+ x y)

5
))
(assert (=

(+ x y)

5))
```

`<leader>F` also gives me a parsing error, but only when the cursor is on the second assertion.

**Comment by bohlender** [comment]
I've confirmed and [fixed the first issue](https://github.com/bohlender/vim-smt2/commit/44bc04969879f238a93450a50b2cf60599ea50aa). The empty line 4 in the S-expression was mistakenly identified as the start of the current paragraph.

I couldn't reproduce the second issue in the latest state. I think [this was fixed yesterday](https://github.com/bohlender/vim-smt2/commit/5bcfd6dc53035aa6e66f7c458d918f10c9f5ddbd) and your plugin is not quite up-to-date.
